test/test_charconv.cpp

```diff
@@ -45,12 +45,12 @@ TEST_CASE("itoa.shortbuf")
                                                             \
         buf0.fill('?');                                     \
         size_t ret0 = fn(buf0, in, rdx);                    \
-        CHECK_EQ(ret0, expected.len);                       \
+        CHECK_GE(ret0, expected.len);                       \
```

Comment:
Those tests are not converting any string anymore, are they ?

I think the modification checking `maxdigits` is a bit intrusive regarding the API.

Maybe it could fallback on old version when test fails ?

Or there may be compromise solutions (maybe first comparing max then using a more accurate computation):
- use msb for binary format (by the way maybe using compilers builtins in memory_util.hpp could help)
- use `(msb / 8 + 1) *2` for hex (mainly bit shifts)
- use `(msb / 3 + 1)` for octal
- would have to scratch my head for dec XD

Reply:
> Those tests are not converting any string anymore, are they ?

Indeed. The implication now is that unless the buffer is "large enough for anything that may appear", the function won't do anything. On the other hand, if the buffer is large enough, then the function efficiently carries to completion and returns the actually used size.

I think this is reasonable and not without precedent. Sort of like a business refusing do to business with another without some capital guarantee insurance being available, even if that guarantee may exceed the amount of business.

> I think the modification checking maxdigits is a bit intrusive regarding the API.

Maybe. But only in a subtle way. The client-side semantics is essentially the same, but now requiring using the latest returned size, whereas previously the returned size would never change:
   * before this PR, the first returned size would be unchanged through repeated calls to `write_dec()` regardless of the input buffer size.
   * with this PR the returned size is clamped at max_digits if the buffer is smaller than that, and then actual digits once the buffer is large enough (so guaranteed smaller than max_digits); 

So with these changes it is still safe to do a speculative call with an empty buffer, then resize it to the returned max_digits size, then call again to have the function actually write to completion. The only catch is that the next returned size is now smaller so you now need to take the latest return.

So the only code that stands to be broken by this change is code like this:
```c++
int val = 0xbeef;
std::vector<char> buf; // start empty
size_t sz = c4::itoa(to_substr(buf), val); // try with empty buffer
buf.resize(sz);
sz = c4::itoa(to_substr(buf), val); // try with large-enough buffer
for(char c : buf) // ERROR: reads uninitialized memory!
   ... // do something.
```
Whereas if the code had played it safe it should be:
```c++
int val = 0xbeef;
std::vector<char> buf; // start empty
size_t sz = c4::itoa(to_substr(buf), val); // try with empty buffer
buf.resize(sz);
sz = c4::itoa(to_substr(buf), val); // try with large-enough buffer
buf.resize(sz); // <--- HERE. always take the latest return value
for(char c : buf) // OK, buf is now trimmed to the written string
   ... // do something.
```
For example, consider [`c4::catrs()`](https://github.com/biojppm/c4core/blob/216b141495e4f8b5bf5ebafbf75b3d7bff583fc9/src/c4/format.hpp#L734); its code works exactly the same with both approaches, because it already used the latest returned size:
```c++
template<class CharOwningContainer, class... Args>
inline void catrs(CharOwningContainer * C4_RESTRICT cont, Args const& C4_RESTRICT ...args)
{
retry:
    substr buf = to_substr(*cont);
    size_t ret = cat(buf, args...);
    cont->resize(ret);
    if(ret > buf.len)
        goto retry;
}
```

Reply:
> Maybe it could fallback on old version when test fails ?
Do you mean do actual digits computation if the input buffer size is smaller than max digits? That may be reasonable, but it would increase the number of instructions so it would have a cost, both in code complexity and maintainability. Is it a steep cost? Maybe, or maybe not, but I'm limited in how much time I can devote to this, and as I've argued above I think this new approach is reasonable. All of the `catrs()`-style code stayed unchanged throughout, which speaks to how similar in intent this new approach is to the previous approach. 

> Or there may be compromise solutions (maybe first comparing max then using a more accurate computation)

Ah, maybe this is reasonable. But it is definitely harder to get right this way, specially with decimals. And it is definitely more expensive vis a vis comparing with maxdigits, which is a compile-time constant.

Reply:
One negative consequence of this changeset I did not address: if you call `catrs()`-style functions with an empty buffer for, say, 100 integers in binary format then the estimated size will be too large, as it is obtained from multiplying the max digits for each integer.

The response is that this is an unlikely scenario. And even so, it is still possible to work around this by providing a likely-enough size. Then the estimation will be more accurate (still larger, but not much larger), because max digits is only returned when the size is not enough.

And ultimately this fits with the `vector<T>::reserve()` mindset. This leaves the client with a choice (and responsibility), where they can (and should) have an idea of the resource requirements for their code. And if they don't have this idea, they have alternative approaches to play it safer, like iterating themselves over the arguments while keeping in control of the buffer size.

Reply:
> by the way maybe using compilers builtins in memory_util.hpp could help

Indeed! I'd like that as well.

Reply:
> > Maybe it could fallback on old version when test fails ?
> 
> Do you mean do actual digits computation if the input buffer size is smaller than max digits? That may be reasonable, but it would increase the number of instructions so it would have a cost, both in code complexity and maintainability. Is it a steep cost? Maybe, or maybe not, but I'm limited in how much time I can devote to this, and as I've argued above I think this new approach is reasonable. All of the `catrs()`-style code stayed unchanged throughout, which speaks to how similar in intent this new approach is to the previous approach.
> 

That would have a cost only when buffer size < max_digits, for the decimals I'd have a couple ideas in mind (making it quite similar to hex).

But your comments seams reasonable.